File: Searching/C++/Aggressive Cows.cpp
class Solution {
  public:
    
    bool check(vector<int> &arr, int k, int dist) {
        int cnt = 1;
        int prev = arr[0];
        
        for (int i = 1; i < arr.size(); i++) {
            if (arr[i] - prev >= dist) {
                prev = arr[i];
                cnt++;
            }
        }
        return (cnt >= k);
    }
    
    int aggressiveCows(vector<int> &arr, int k) {
        // code here
        sort(arr.begin(), arr.end());
        int res= 0;
        int lo = 1;
        int hi = arr.back() - arr[0];
        
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            
            if (check(arr, k, mid)) {
                res = mid;
                lo = mid + 1;
            }
            else {
                hi = mid - 1;
            }
        }
        return res;
    }
};

File: Searching/Python/Aggressive Cows.py
class Solution:
    def check(self, arr, k, dist):
        cnt = 1
        prev = arr[0]
        
        for i in range(1, len(arr)):
            if arr[i] - prev >= dist:
                prev = arr[i]
                cnt += 1
        return cnt >= k
    
    def aggressiveCows(self, arr, k):
        # code here
        arr.sort()
        res = 0
        
        lo = 1
        hi = arr[-1] - arr[0]
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            
            if self.check(arr, k, mid):
                res = mid
                lo = mid + 1
            else:
                hi = mid - 1
        return res
        
